File: src/utils/generateData_findTest.ts
const femaleCharacters = {
    Angelagbaby: 8,
    迪丽热巴: 36,
    范冰冰: 9,
    杨幂: 46,
    赵丽颖: 9,
};

const maleCharacters = {
    李晨: 12,
    鹿晗: 16,
    郑恺: 20,
};

let ID = 0;

function generateId() {
    ID++;
    return ID.toString();
}

export function getData() {
    const characters = Math.random() < 0.5 ? femaleCharacters : maleCharacters;
    const person1 = randomChoice(Object.keys(characters));
    const data = {};
    const imagesNumber1 = characters[person1];
    const imagesIndex1 = randomSample(
        [...Array(imagesNumber1).keys()].map((i) => i + 1),
        2,
    );
    data['target'] = {
        id: generateId(),
        image: `public/face/${person1}/${imagesIndex1[0]}.jpg`,
        name: person1,
    };
    const tests = [];
    const images = [];
    for (let i = 0; i < 7; i++) {
        let newPerson = randomChoice(Object.keys(characters));
        while (newPerson === person1) {
            newPerson = randomChoice(Object.keys(characters));
        }
        let imageIndex = `public/face/${newPerson}/${randomChoice([...Array(characters[newPerson]).keys()].map((i) => i + 1))}.jpg`;
        while (images.includes(imageIndex)) {
            imageIndex = `public/face/${newPerson}/${randomChoice([...Array(characters[newPerson]).keys()].map((i) => i + 1))}.jpg`;
        }
        images.push(imageIndex);
        tests.push({
            id: generateId(),
            image: imageIndex,
            name: newPerson,
            answer: false,
        });
    }
    tests.push({
        id: generateId(),
        image: `public/face/${person1}/${imagesIndex1[1]}.jpg`,
        name: person1,
        answer: true,
    });
    shuffle(tests);
    data['tests'] = tests;
    return data;
}

function randomChoice(arr) {
    return arr[Math.floor(Math.random() * arr.length)];
}

function randomSample(arr, size) {
    const result = [];
    const set = new Set();
    while (set.size < size) {
        const index = Math.floor(Math.random() * arr.length);
        if (!set.has(index)) {
            set.add(index);
            result.push(arr[index]);
        }
    }
    return result;
}

function shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}

File: public/generateData_findTest.py
import json
import random

female_characters = {
    "Angelagbaby": 8,
    "迪丽热巴": 36,
    "范冰冰": 9,
    "杨幂": 46,
    "赵丽颖": 9,
}
male_characters = {
    "李晨": 12,
    "鹿晗": 16,
    "郑恺": 20,
}

ID = 0


def generate_id():
    global ID
    ID += 1
    return str(ID)


def get_data():
    global ID
    characters = random.choice([female_characters, male_characters])
    person1 = random.choice(list(characters.keys()))
    data = {}
    images_number_1 = characters[person1]
    images_index_1 = random.sample(range(1, images_number_1 + 1), 2)
    data['target'] = {
        'id': generate_id(),
        'image': f'public/face/{person1}/{images_index_1[0]}.jpg',
        'name': person1
    }
    tests = []
    images = []
    for i in range(9):
        # 随机抽一个新的人物
        new_person = random.choice(list(characters.keys()))
        while (new_person == person1):
            new_person = random.choice(list(characters.keys()))
        # 生成图片地址
        image_index = f'public/face/{new_person}/{random.choice(range(1, characters[new_person] + 1))}.jpg'
        if (image_index in images):
            while (image_index in images):
                image_index = f'public/face/{new_person}/{random.choice(range(1, characters[new_person] + 1))}.jpg'
        # 存储新生成的数据
        images.append(image_index)
        tests.append({
            "id": generate_id(),
            'image': image_index,
            'name': new_person,
            'answer': False
        })
    tests.append({
        "id": generate_id(),
        'image': f'public/face/{person1}/{images_index_1[1]}.jpg',
        'name': person1,
        'answer': True
    })
    # 打乱数组
    random.shuffle(tests)
    data['tests'] = tests

    return data


def main():
    global ID
    ID = 0
    data = get_data()
    json_data = json.dumps(data, ensure_ascii=False, indent=4)
    print(json_data)


if __name__ == '__main__':
    main()
